Pulsar connector (#2002)
smallrye/smallrye-reactive-messaging: documentation/src/main/docs/pulsar/receiving-pulsar-messages.md

# Receiving messages from Pulsar

The Pulsar Connector connects to a Pulsar broker using a Pulsar client and creates consumers to
receive messages from Pulsar brokers and it maps each of them into Reactive Messaging `Messages`.

## Example

Let’s imagine you have a Pulsar broker running, and accessible
using the `pulsar:6650` address (by default it would use
`localhost:6650`). Configure your application to receive Pulsar messages
on the `prices` channel as follows:

```properties
mp.messaging.incoming.prices.connector=smallrye-pulsar # <1>
mp.messaging.incoming.prices.serviceUrl=pulsar://pulsar:6650 # <2>
mp.messaging.incoming.prices.schema=DOUBLE # <3>
mp.messaging.incoming.prices.subscriptionInitialPosition=Earliest # <4>
```

1.  Sets the connector for the `prices` channel

> 

Reviewer: you sure about the extra blank line here?

```

## Consumer Configuration

The Pulsar Connector allows flexibly configuring the underlying Pulsar consumer.
One of the ways is to set consumer properties directly on the channel configuration.
The list of available configuration properties are listed in [Configuration Reference](#configuration-reference).

See the [Configuring Pulsar consumers, producers and clients](client-configuration.md) for more information.

## Deserialization and Pulsar Schema

The Pulsar Connector allows configuring Schema configuration for the underlying Pulsar consumer.
See the [Configuring the schema used for Pulsar channels](schema-configuration.md) for more information.

## Inbound Metadata

The incoming Pulsar messages include an instance of `PulsarIncomingMessageMetadata` in the metadata.
It provides the key, topic, partitions, headers and so on:

```java
{{ insert('pulsar/inbound/PulsarMetadataExample.java', 'code') }}
```

## Acknowledgement

When a message produced from a Pulsar Message is *acknowledged*, the connector sends an [acknowledgement request](https://pulsar.apache.org/docs/3.0.x/concepts-messaging/#acknowledgment) to the Pulsar broker.

> All Reactive Messaging messages needs to be *acknowledged*, and in most cases this is handled automatically.

Reviewer: ```suggestion
All Reactive Messaging messages need to be *acknowledged*, which is handled automatically in most cases.
```
